Amqp/src/Akka.Streams.Amqp/AmqpConnector.cs

```diff
@@ -0,0 +1,178 @@
+                    if (details.ConnectionTimeout.HasValue)
+                        factory.ContinuationTimeout = details.ConnectionTimeout.Value;
+                    if (details.HandshakeTimeout.HasValue)
+                        factory.HandshakeContinuationTimeout = details.HandshakeTimeout.Value;
+                    break;
+                }
+                case DefaultAmqpConnection defaultConnection:
+                    //leave it be as is
+                    break;
+            }
+            return factory;
+        }
+
+        public static IConnection NewConnection(IConnectionFactory factory, IAmqpConnectionSettings settings)
+        {
+            switch (settings)
+            {
+                case AmqpConnectionDetails details:
+                {
+                    if (details.HostAndPortList.Count > 0)
+                    {
+                        return factory.CreateConnection(details.HostAndPortList
+                            .Select(pair => new AmqpTcpEndpoint(pair.host, pair.port)).ToList());
+                    }
+                    else
+                    {
+                        throw new ArgumentException("You need to supply at least one host/port pair.");
+                    }
+                }
+                default:
+                    return factory.CreateConnection();
+            }
+        }
+    }
+
+    /// <summary>
+    /// Internal API
+    /// </summary>
+    internal abstract class AmqpConnectorLogic : GraphStageLogic
+    {
+        protected IConnection Connection;
+        protected IModel Channel;
+        protected Action<ShutdownEventArgs> ShutdownCallback;
+
+        protected AmqpConnectorLogic(Shape shape) 
+            : base(shape)
+        {
+            
+        }
+
+        public abstract IAmqpConnectorSettings Settings { get; }
+
+        public abstract IConnectionFactory ConnectionFactoryFrom(IAmqpConnectionSettings settings);
+
+        public abstract IConnection NewConnection(IConnectionFactory factory, IAmqpConnectionSettings settings);
+
+        public abstract void WhenConnected();
+
+        public abstract void OnFailure(Exception ex);
+
+        public override void PreStart()
+        {
+            try
+            {
+                var factory = ConnectionFactoryFrom(Settings.ConnectionSettings);
+                Connection = NewConnection(factory, Settings.ConnectionSettings);
+                Channel = Connection.CreateModel();
+                ShutdownCallback = GetAsyncCallback<ShutdownEventArgs>(args =>
+                {
+                    if (args.Initiator != ShutdownInitiator.Application)
+                        FailStage(ShutdownSignalException.FromArgs(args));
+                });
+                Connection.ConnectionShutdown += OnConnectionShutdown;
+                Channel.ModelShutdown += OnChannelShutdown;
+
+                Settings.Declarations.ForEach(declaration =>
```

Comment:
I don't know how about others: I prefer foreach loop - it doesn't allocate delegates and gives the space for potential inlining.